File: 15-3sum/15-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
  if (nums.length < 3) return []
  nums.sort((a, b) => a - b)
  const len = nums.length
  if (nums[0] > 0 || nums[len-1] < 0) return []
  const twoSum = (i, target) => {
    const result = []
    let l = i
    let r = len - 1
    while (l < r) {
      const left = nums[l]
      const right = nums[r]
      const sum = left + right
      if (sum === target) {
        result.push([left, right])
        l ++
        r --
        while (nums[l] === left) l++
        while (nums[r] === right) r--
      } else if (sum < target) {
        l ++
        while (l < r && nums[l] === left) l++
      } else {// sum > target
        r --
        while (l < r && nums[r] === right) r--
      }
    }
    return result    
  }
  const ans = []
  for (let i = 0; i < len - 2;) {
    const a = nums[i]
    if (a > 0) break
    const res = twoSum(i+1, -a)
    if (res.length > 0) {
      const temp = res.map(arr => [a, ...arr])
      ans.push(...temp)
    }
    i++
    while (i < len - 2 && nums[i] === a) i++
  }
  return ans
};

File: 15-3sum/15-3sum.py
class Solution:
  def threeSum(self, nums: List[int]) -> List[List[int]]:
    ln = len(nums)
    if ln < 3: return []
    nums.sort()
    if nums[0] > 0 or nums[-1] < 0: return []
    if nums[0] == 0 and nums[-1] == 0: return [[0, 0, 0]]
    result = []
    i = 0
    while i < ln-2:
      a = nums[i]
      if a > 0: return result
      l = i + 1
      r = ln - 1
      while l < r:
        b = nums[l]
        c = nums[r]
        if c < 0: return result
        if a + b + c < 0:
          l += 1
          while l < r and nums[l] == b: l += 1
        elif a + b + c == 0:
          result.append([a, b, c])
          l += 1
          r -= 1
          while l < r and nums[l] == b: l += 1
          while l < r and nums[r] == c: r -= 1
        else:
          r -= 1
          while l < r and nums[r] == c: r -= 1
      i += 1
      while i < ln-2 and nums[i] == a: i += 1

    return result